### PR TITLE
HDDS-1812. Du while calculating used disk space reports that chunk files are file not found

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

1. Copy `DU` from `hadoop-common` and customize it to ignore temporary chunk files.  Unfortunately the original implementation could not be simply extended.  Note that this means temp chunk files will not be counted towards HDDS space usage.  One possible alternative is ignoring errors from `du` (which returns with error due to the missing files), but that might make it more difficult to spot other problems.
2. Introduce an interface for creating various "space usage" implementations.  `hadoop-common` allows a single implementation to be configured, which is used for all volumes.  The factory interface allows creating different implementations or different parameters for different volumes.  Currently 3 factories are implemented:
   1. pure `du`-based
   2. pure `df`-based (this is actually based on `java.io` in both Hadoop Common and HDDS)
   3. mock for testing -- this also helps avoid test failures due to "disk out of space" errors when allocating containers
   Hybrid or other more intelligent implementation could be a future improvement.  Eg. one that would use `du` for shared volumes and `df` for dedicated ones.
3. Introduce an interface for the object that persists space usage info (to a file named `scmUsed`) to be able to:
   1. skip persisting info from cheaper sources (ie. `df`)
   2. use in-memory persistence for unit testing
4. Decouple caching and periodic refresh of space usage from its source.  Use a `ScheduledExecutorService` for refreshing disk usage info.  Use a "same thread executor" for unit tests.
5. Let `container-service` tests use code from HDDS `common` tests, which provide mock implementations of the new interfaces.  This also required:
   1. Avoid using real config properties (eg. `ozone.scm.client.bind.host`) for testing config file generation.  The generated config file is picked up by other tests and causes failures.
   2. Rename Log4J2 config file used by audit logger test to avoid creating an untracked `audit.log` (similar to HDDS-2063).  This would happen if some test starts components which use audit logger, and it picks up `log4j2.properties` by default.

https://issues.apache.org/jira/browse/HDDS-1812

## How was this patch tested?

Added several unit tests, even for logic extracted from existing classes.  On compose-based cluster tested configuring each of the real factory implementations and configuring different refresh periods.  Tested `du`-based implementation by continuously creating/renaming files from shell.

### CODE REVIEW

**Comment by adoroszlai** [comment]
Thanks @elek for the review and offline discussion.

I updated the patch to use `@Config` objects for the factories (both for specific implementations and the global one that creates it).  I agree that this makes the code more encapsulated, instead of scattering config info around several files (`ozone-default.xml`, `HddsConfigKey`, etc.)

I also renamed the factory config property a bit (`...class` to `...classname`, since it's a string).

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/AbstractSpaceUsageSource.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  protected AbstractSpaceUsageSource(File file) {
+    this.file = file;
+    try {
+      this.path = file.getCanonicalPath();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  protected static long time(LongSupplier supplier, Logger logger) {
```
Comment (arp7) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/AbstractSpaceUsageSource.java]: Could you add a short javadoc? I got what it does but not clear from first reading.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+          running = true;
+        }
+      }
+    } else {
+      refresh();
+    }
+  }
+
+  public void shutdown() {
+    synchronized (this) {
```
Comment (arp7) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java]: I don't think we need the synchronization. Looks like we only synchronize in stop and start, and user of the class should ensure they don't invoke both concurrently.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/AbstractSpaceUsageSource.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/AbstractSpaceUsageSource.java]: Good idea, I had to re-read it, too. ;)

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java (hunk 2 above)
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java]: Good point, `VolumeUsage` already synchronizes start and shutdown.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDU.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  @Test
+  public void testExcludePattern() throws IOException {
+    createFile(new File(DIR, "include.txt"), (int) (4 * KB));
+    createFile(new File(DIR, "exclude.tmp"), (int) (100 * KB));
+    SpaceUsageSource du = new DU(DIR, "*.tmp");
+
+    long usedSpace = du.getUsedSpace();
+
+    assertFileSize(KB, usedSpace);
```
Comment (hanishakoneru) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDU.java]: Expected size here should be 4*KB.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDedicatedDiskSpaceUsage.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    FileUtil.fullyDelete(DIR);
+  }
+
+  @Test
+  public void testGetUsed() throws IOException {
+    File file = new File(DIR, "data");
+    createFile(file, FILE_SIZE);
+    SpaceUsageSource subject = new DedicatedDiskSpaceUsage(DIR);
+
+    assertTrue(subject.getUsedSpace() >= FILE_SIZE - 20);
```
Comment (hanishakoneru) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDedicatedDiskSpaceUsage.java]: Can you please add a comment explaining -20 for margin of error.

Comment (adoroszlai) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDedicatedDiskSpaceUsage.java]: It comes from [`TestDFCachingGetSpaceUsed`](https://github.com/apache/hadoop/blob/2aa065d98f36527d7769c9c58a923a706036391d/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestDFCachingGetSpaceUsed.java#L61) from Hadoop. :)

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDU.java (hunk 3 above)
Comment (adoroszlai) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/fs/TestDU.java]: Thanks, updated.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+  public CachingSpaceUsageSource(SpaceUsageCheckParams params,
+      ScheduledExecutorService executor) {
+    Preconditions.checkArgument(params != null, "params == null");
+
+    refresh = params.getRefresh();
+    source = params.getSource();
+    persistence = params.getPersistence();
+    this.executor = refresh.isZero() ? null : executor;
+
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java]: Replaced with a precondition to ensure that "zero refresh" and "null executor" are either both true or both false.  Also adjusted the unit test to match this condition.

**Comment by adoroszlai** [comment]
Thanks @hanishakoneru for the kind review.  Updated the patch based on your comments.

**Comment by hanishakoneru** [comment]
Thank you @adoroszlai. I am +1 for merge.

**Pull request merged**